### ISSUE TITLE
LG-330: MicroProfile Metrics v5.0 Implementation

### ISSUE THREAD

**Issue by donbourne** [OpenLiberty/open-liberty#20018]
## Description of the high level feature, including any external spec links:  

Implement MicroProfile Metrics v5.0 in Open Liberty

- [ ] inform MicroProfile FT team when it's time to test FT with early driver (see #22046)
- [ ] create standalone blog explaining new capabilities and migration tips
- [ ] update open liberty guide (https://openliberty.io/guides/microprofile-metrics.html#adding-microprofile-metrics-to-the-inventory-service)
- [ ] update open liberty operator documentation for monitoring
- [ ] update grafana dashboards
- [ ] provide names of new open source, the versions, the licenses, and when they will be going in to @CassTucker to get clearance

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## When complete & mandatory, add links to the UFO (Upcoming Feature Overview) document, FTS (Feature Test Summary), blogs post issues(s), and [Aha](https://cloud-platform.ideas.aha.io/) (externally raised RFEs):
- UFO: https://ibm.box.com/s/z5pbrb114ge81m2pfjzsntt4v394zqwt 
- FTS:
- Beta Blog Post (if applicable):
- Blog Post: GA BLOG - MicroProfile Metrics 5.0 [#24596](https://github.com/OpenLiberty/open-liberty/issues/24596)
- Aha: 
   - Also, add a comment into the Aha idea linking to this issue.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [ ] No Design / No UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
  * Note: Design is required if any of the following apply: SVT or Performance testing, new UI, Servicibility considerations, significant documentation changes
- [x] ID Requirements identified ([Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty)). (Epic owner / Feature owner with ID focal point)
  * Note: If only trivial documentation changes are required, you may reach out to the ID Feature Focal to request a `ID Required - Trivial` label.  Unlike features with regular ID requirement, those with `ID Required - Trivial` label do not have a hard requirement for a Design/UFO.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] Create a child task of this epic entitled "Feature Test Summary" via [this template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=). Add the link in above.

##
##

## **Development**
When active development has begun 
- [x] Add the "In Progress" label to this issue.

## **Beta**
If your feature, or portions of it, are going to be included in a beta  
Before Onboarding the beta
- [ ] Beta Fence the functionality (`kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`)  

1 week before beta GA
- [x] Create, populate, and link to the [Beta blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_beta.md&title=BETA+BLOG+-+title_of_your_update)

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - ([Karen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
  * Note: If only trivial documentation changes are required, you may reach out to the ID Feature Focal to request a `ID Required - Trivial` label.  Unlike features with regular ID requirement, those with `ID Required - Trivial` label do not have a hard requirement for a Design/UFO.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the release branch. (Epic owner / Feature owner / Backlog Subtribe PM)

1 week before GA
- [ ] Create, populate, and link to the [Blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_ga_release.md&title=GA+BLOG+-+title_of_your_update)

## **Other deliverbles**
- [ ] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [ ] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by donbourne** [comment]
discussed need for perf testing with @jdmcclur .  Plan is to get started on that once we have code merged - sometime around Sept. 6.

**Comment by chirp1** [comment]
Content received to document for the doc issue at https://github.com/OpenLiberty/docs/issues/5814 . A draft has been written. Approving this epic.

**Comment by tngiang73** [comment]
STE slides: https://ibm.box.com/s/iizlvio8wpru9ffzlmvidc7q8kejrh8x

**Comment by Channyboy** [comment]
@donbourne 
Serviceability Approval Comment - Please answer the following questions for serviceability approval:

UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?

_**Yes (slide 35).
We anticipate errors with users configurating external micrometer libraries with the mpMetrics feature. Most likely scenario is that the customer does not provide the right dependencies for their micrometer registry jars or even forget to provide the micrometer core jar.
If the Micrometer Core is missing, an error is emitted as it is critical for the mpMetrics feature (i.e. the underlying SmallRye implementation) to have the Micrometer Core classes to to function.**_

_**Otherwise, if if dependencies for a Micrometer meter registry are missing, since it is a plug and play mechanics the Micrometer framework that the SmallRye implementation relies on, we can not foresee any missing dependencies until a runtime exception occurs due to `ClassNotFoundException` or `NoClassDefError` is witnessed. If it is detected during feature startup, we will emit an error message along side an accompanying FFDC. Otherwise an FFDC occurs during runtime when the Micrometer code is executing.**_

_**Another anticipated issue would the the enabling and configuration of the micrometer registry to connect to the desired monitoring system. These are done through the MP Config properties.
There are OL documentation for this (which largely redirects to Micrometer documentation). L2 is also briefed about this through the STE documentation provided to them.**_


Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).

a) What problem paths were tested and demonstrated?
- Missing dependency for a Micrometer Meter Registry in a BYOL scenario for mpMetrics
- Error thrown by SmallRye when starting mpMetrics with a BYOL scenario and a Prometheus Meter Registry is not provided.
- Missing configuration/mis-configured property for a Micrometer meter registry.

b) Who did you demo to?

_**WAS L3: Liberty Log Analytics (other members of the team did nor participate in the development)**_

c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

_**As the mechanism for the "bring your own (micrometer) library" depends on the Micrometer framework, and the fact that the implementation is unable to foresee any missing dependencies or missing/mis-configured properties, there is concern whether the Micrometer libraries themselves provide adequate logging to help the customer resolve the problem.**_

_**The documentation we provide, plus briefing to the L2, stresses that customer should evaluate the Micrometer documentation to ascertain the appropriate dependencies and configure the properties correctly. As it is a third-party framework that we are leveraging for this functionality to be successful, we are ultimately not directly responsible.**_

_**Action to take was to make the OL documentation for the libraryRef configuration use a more complicated meter registry as an example to better walkthrough how to derive the registry and property values from the Micrometer documentation.**_

SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?

_**Daniel Guinan**_

b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?

_**yes**_

c)Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.

_**WAS L3: Liberty Log Analytics**_

d)Does this feature add any new metrics or emit any new JSON events? If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

_**No new metrics or events.**_

**Comment by donbourne** [comment]
@pgunapal , can you please do the serviceability review/approval on this one?  I was involved in the UFO and other parts, so it's not appropriate for me to do the approval.
